### DISCUSSION TITLE
IP Merge Fails in def2stream.py

### DISCUSSION THREAD

**Discussion by CathedralsOfSand** [The-OpenROAD-Project/OpenROAD-flow-scripts#1993] [Q&A] [answered]
I’ve run designs gcd, ibex on nangate45, sky130hd and asap7 and all of them fail during the main_layout.read(def_in, layoutOptions) step. If I remove the layoutOptions object it runs but leaves std cell internal layers empty.  IP merge is a fundamental step to chip finish.  Any help is appreciated.

**Comment by maliberty** [comment]
What version of klayout are you using?  Did you use the dependency installer?

**Reply by CathedralsOfSand** [reply]
Yes I managed to manually install  the dependencies. The challenges is OpenRoad only supports Debian11 and Klayout only supports version 0.26.2 on Debian11. To get to 0.28.15 I need Debian12 but OpenRoad does not support Debian12.

**Comment by CathedralsOfSand** [comment]
0.26.2. Where can I find the dependency installer?

When layoutOptions is used it can’t read the file because it put the final.def on the same line as the tech.lef from the .lyt file.

**Comment by CathedralsOfSand** [comment]
It tells me Debian is an unsupported system

**Comment by CathedralsOfSand** [comment]
I installed the packages list in Dependency.sh for Debian. No change.

**Comment by maliberty** [comment, marked as answer]
0.26.2 is too old.  We are currently using 0.28.8 but I think you need at least 0.27.

**Reply by CathedralsOfSand** [reply]
Matt how do I resolve the version conflict between openroad, Klayout and Linux?

Ubuntu 24 supports Klayout 28 but OpenRoad only supports Ubuntu 22 which only supports klayout 26. Similar story for Debian.

**Reply by maliberty** [reply]
The dependency installer works on Ubuntu22 and installs klayout.  See https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/3d347b425f4f8ecc1e514e2b6504f96592fabd53/etc/DependencyInstaller.sh#L99

**Reply by CathedralsOfSand** [reply]
Thanks Matt. I’m on Debian11. Tomorrow I will start over on an Ubuntu 22 virtual machine.

**Reply by CathedralsOfSand** [reply]
I’ll let you know how it goes. Hopefully not the same outcome.

**Reply by maliberty** [reply]
We didn't officially support Debian.  Even on Ubuntu we directly install layout and didn't rely on the package managers version

**Reply by CathedralsOfSand** [reply]
Thanks for the information. I also had to edit the following line in the MakeFile because it got a syntax error:

# Global setting for Synthesis
export SYNTH_ARGS ?= -flatten -extra-map $(FLOW_HOME)/platforms/common/lcu_kogge_stone.v

yosys recognizes the -flatten argument but not -extra-map

I believe that should be a separate extract -map command. Is this a yosys version issue or a MakeFile issue?

**Reply by maliberty** [reply]
yosys but we include yosys a part of ORFS so you should be getting that version and not some system version which might be older.

**Reply by CathedralsOfSand** [reply]
Matt thanks again for your support. I get an error when I execute the Dependency.sh

sudo sh Dependency.sh
3: set: Illegal option -o pipefail 

if I remove pipefail I get
7: OSTYPE: parameter not set

**Reply by maliberty** [reply]
Per https://openroad-flow-scripts.readthedocs.io/en/latest/user/BuildLocally.html you should run 
```
git clone --recursive https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts
cd OpenROAD-flow-scripts
sudo ./setup.sh
```

**Reply by CathedralsOfSand** [reply]
My bad. I was trying to build from binaries.

**Comment by CathedralsOfSand** [comment]
Thank you Matt. Appreciate your support. I will update. I am new to OpenRoad but from my experience with commercial tools this looks like a flow issue. I could be wrong. When layoutOptions is included in def2stream.py

main_layout.read(in_def, layoutOptions)

The full path to def_in is concatenated with the tech.lef relative path in one string and I get an error reading the file. I found the path to the tech.lef in the technology file. The string looks something like 

fullpath/6_final.def/../../../../objects/platform/design/tech.lef

If I remove layoutOptions and just use main_layout.read(in_def), the flow executes to completion but only the top level wires are in the final.gds.

I can manually import the final.def, the tech.lef and the standard cell.lef into klayout and save the gds which gets a much more complete gds but still the standard cell internal layers are missing (e.g. no poly, contact, diffusion etc...). I tried the merge hierarchy option but it didn't work.

This suggests there is a flow issue but likely dealing with two separate issues: layoutOptions and IP merge.

**Comment by CathedralsOfSand** [comment]
Thanks Matt gcd nangate45 ran completely. All layers visible. Issue resolved. I will file a separate issue for the -extra-map. That still fails. And there was a DFF_LIB_FILE issue with flop mapping. But worked around it.
